format-avro/src/main/java/io/cdap/plugin/format/avro/output/DelegatingAvroKeyRecordWriter.java

```diff
@@ -0,0 +1,87 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.avro.output;
+
+import org.apache.avro.Schema;
+import org.apache.avro.file.CodecFactory;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.hadoop.io.AvroSerialization;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.avro.mapreduce.AvroJob;
+import org.apache.avro.mapreduce.AvroKeyRecordWriter;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * TODO: add
+ */
+public class DelegatingAvroKeyRecordWriter extends RecordWriter<AvroKey<GenericRecordWrapper>, NullWritable> {
+  private final Map<Integer, RecordWriter<AvroKey<GenericRecord>, NullWritable>> delegateMap;
+  private final TaskAttemptContext context;
+  private final CodecFactory codecFactory;
+  private final OutputStream outputStream;
+  private final int syncInterval;
+
+  public DelegatingAvroKeyRecordWriter(TaskAttemptContext context,
+                                       CodecFactory codecFactory,
+                                       OutputStream outputStream,
+                                       int syncInterval) {
+    super();
+    this.delegateMap = new HashMap<>();
+    this.context = context;
+    this.codecFactory = codecFactory;
+    this.outputStream = outputStream;
+    this.syncInterval = syncInterval;
+  }
+
+  @Override
+  public void write(AvroKey<GenericRecordWrapper> key, NullWritable value) throws IOException, InterruptedException {
+    RecordWriter<AvroKey<GenericRecord>, NullWritable> delegate =
+      delegateMap.computeIfAbsent(key.datum().getHash(), (k) -> {
+        try {
+          Configuration conf = context.getConfiguration();
+          Schema writerSchema = key.datum().getSchema();
+          GenericData dataModel = AvroSerialization.createDataModel(conf);
+
+          return new AvroKeyRecordWriter<>(writerSchema,
+                                           dataModel,
+                                           codecFactory,
+                                           outputStream,
+                                           syncInterval);
+        } catch (IOException ioe) {
+          throw new RuntimeException("Unable to initialize Avro Record Writer", ioe);
+        }
+      });
+
+    delegate.write(new AvroKey<>(key.datum().getGenericRecord()), value);
+  }
+
+  @Override
+  public void close(TaskAttemptContext context) throws IOException, InterruptedException {
+    for (RecordWriter<AvroKey<GenericRecord>, NullWritable> delegate : delegateMap.values()) {
+      delegate.close(context);
```

Comment:
if one of these throws IOException, should still try to close the others before eventually throwing an exception.

If more than one throws an exception, can use .addSuppressed() to make sure all exceptions are captured in whatever is eventually thrown.

Reply:
I have implemented this logic.